Engine/src/Engine/InputSystem.h

```diff
@@ -0,0 +1,120 @@
+#pragma once
+#include <map>
+#include <mutex>
+#include "glm/vec2.hpp"
+
+struct GLFWwindow;
+
+namespace Engine 
+{
+	/// <summary>
+	/// Provides a wrapper for GLFW input
+	/// </summary>
+	class InputSystem
+	{
+	public:
+		//Delete copy c'tor and assignment opertor for singleton pattern
+		InputSystem(InputSystem& other) = delete;
+		void operator=(InputSystem& other) = delete;
+
+#pragma region Public Methods
+		/// <summary>
+		/// Initialize the input system
+		/// </summary>
+		/// <param name="window">Window context</param>
+		void init(GLFWwindow* window);
+		/// <summary>
+		/// Checks for a mouse button press
+		/// </summary>
+		/// <param name="button">Mouse button to check</param>
+		/// <returns>If the button has been pressed</returns>
+		bool isButtonPressed(int button);
+		/// <summary>
+		/// Checks for a key press
+		/// </summary>
+		/// <param name="key">Key to check</param>
+		/// <returns>If the key has been pressed</returns>
+		bool isKeyPressed(int key);
+		/// <summary>
+		/// Set the current window for context, resets initialization to false
+		/// </summary>
+		/// <param name="window">The active window</param>
+		void setWindow(GLFWwindow* window);
+		/// <summary>
+		/// Whether the InputSystem is initialized
+		/// </summary>
+		/// <returns>State of input system</returns>
+		bool isInit();
+		/// <summary>
+		/// Getter for current position of the cursor
+		/// </summary>
+		/// <returns>The current position of the cursor if initialized</returns>
+		glm::vec2& getCursorPos();
+		/// <summary>
+		/// Grabs the singleton instance
+		/// </summary>
+		/// <returns>The instance</returns>
+		static InputSystem* getInstance();
+#pragma endregion
+#pragma region Callbacks
+		/// <summary>
+		/// Callback function registered with glfw
+		/// </summary>
+		/// <param name="window">Window context</param>
+		/// <param name="button">Active button</param>
+		/// <param name="action">Action of button</param>
+		/// <param name="mods">Modifier bits (shift,ctrl,etc.)</param>
+		void mouseCallback(GLFWwindow* window, int button, int action, int mods);

```

Comment:
Is 'Action of Button' just representative of whether or not the button is pressed, released, etc?

Reply:
Yes